### PR TITLE
Middle out

### LINKED ISSUE #774: authorship_api 404s for known WOSUIDs
Repeated example from `production.log`:
```
Started POST "/authorship.json" for 10.35.49.120 at 2018-03-06 16:45:34 -0800
POST Contribution JSON: 
{"cap_profile_id":"35324","featured":false,"status":"approved","visibility":"private","wos_uid":"WOS:000364146300042"}
The WOS:000364146300042 publication was not found either locally or at WebOfScience.
Started POST "/authorship.json" for 10.35.49.120 at 2018-03-06 16:45:36 -0800
POST Contribution JSON: 
{"cap_profile_id":"35324","featured":false,"status":"approved","visibility":"private","wos_uid":"WOS:000364146300042"}
The WOS:000364146300042 publication was not found either locally or at WebOfScience.
Started POST "/authorship.json" for 10.35.49.120 at 2018-03-06 16:45:38 -0800
POST Contribution JSON: 
{"cap_profile_id":"35324","featured":false,"status":"approved","visibility":"private","wos_uid":"WOS:000364146300042"}
The WOS:000364146300042 publication was not found either locally or at WebOfScience.
Started POST "/authorship.json" for 10.35.49.120 at 2018-03-06 16:45:40 -0800
POST Contribution JSON: 
{"cap_profile_id":"35324","featured":false,"status":"approved","visibility":"private","wos_uid":"WOS:000364146300042"}
The WOS:000364146300042 publication was not found either locally or at WebOfScience.
Started POST "/authorship.json" for 10.35.49.120 at 2018-03-06 16:45:44 -0800
POST Contribution JSON: 
{"cap_profile_id":"35324","featured":false,"status":"approved","visibility":"private","wos_uid":"WOS:000364146300042"}
The WOS:000364146300042 publication was not found either locally or at WebOfScience.
```

Corresponding lines from `web_of_science.log` don't seem to suggest any problem:
```
I, [2018-03-06T16:45:34.980010 #10247]  INFO -- : WebOfScience::Harvester - author 17853 -  - 1 UIDs for search
I, [2018-03-06T16:45:34.982041 #10247]  INFO -- : WebOfScience::Harvester - author 17853 -  - 1 UIDs without contributions
I, [2018-03-06T16:45:36.699597 #10247]  INFO -- : WebOfScience::Harvester - author 17853 -  - 1 UIDs for search
I, [2018-03-06T16:45:36.701875 #10247]  INFO -- : WebOfScience::Harvester - author 17853 -  - 1 UIDs without contributions
I, [2018-03-06T16:45:38.336665 #10247]  INFO -- : WebOfScience::Harvester - author 17853 -  - 1 UIDs for search
I, [2018-03-06T16:45:38.338826 #10247]  INFO -- : WebOfScience::Harvester - author 17853 -  - 1 UIDs without contributions
I, [2018-03-06T16:45:40.230009 #10247]  INFO -- : WebOfScience::Harvester - author 17853 -  - 1 UIDs for search
I, [2018-03-06T16:45:40.232269 #10247]  INFO -- : WebOfScience::Harvester - author 17853 -  - 1 UIDs without contributions
I, [2018-03-06T16:45:44.330771 #10247]  INFO -- : WebOfScience::Harvester - author 17853 -  - 1 UIDs for search
I, [2018-03-06T16:45:44.332758 #10247]  INFO -- : WebOfScience::Harvester - author 17853 -  - 1 UIDs without contributions
```

@peetucket, author `17853` is your profile (w/ `cap_profile_id: 35324`) so those match.

### PR DESCRIPTION
**Pull request by atz**
This was a bit of a journey with a lot of housekeeping, so I'm totally open to parting it out to multiple PRs and/or squashing commits.

Fix #774.

### CODE REVIEW

**Comment by peetucket** [comment]
I hope this a Silicon Valley tv show reference

**Comment by atz** [comment]
This was a bit of a journey with a lot of housekeeping, so I'm totally open to parting it out to multiple PRs and/or squashing commits.

Fix #774.

File: lib/web_of_science/process_records.rb [hunk 1]
```diff
@@ -33,49 +41,68 @@ def execute
...
+      # @return [Array<String>] WosUIDs that successfully create a new Publication
       def create_publications
         return [] if records.empty?
-        matching_uids = Publication.where(wos_uid: records.map(&:uid)).pluck(:wos_uid)
-        save_wos_records(records.reject { |rec| matching_uids.include? rec.uid })
-        records.select { |rec| !matching_contribution(author, rec) && create_publication(rec) }
-               .map(&:uid)
-               .uniq
+        wssrs = save_wos_records
+        wssrs_hash = wssrs.map(&:uid).zip(wssrs).to_h
```
Comment (peetucket) [inline lib/web_of_science/process_records.rb]: not sure i understand what is happening here - would you mind just walking me through the logic (in person/zoom if more efficient)?

File: app/api/sul_bib/authorship_api.rb [hunk 2]
```diff
@@ -113,11 +114,8 @@ def get_publication_via_sciencewire!(sw_id)
       # @param [String] wos_uid WebOfScience ID
       # @return [Publication]
       def get_publication_via_wos!(author, wos_uid)
-        pub = Publication.find_by(wos_uid: wos_uid)
-        return pub if pub
-        uids = WebOfScience.harvester.process_uids(author, [wos_uid])
-        uids.present? || log_and_error!("The #{wos_uid} publication was not found either locally or at WebOfScience.")
-        Publication.find_by(wos_uid: uids.first)
+        WebOfScience.harvester.author_uid(author, wos_uid) ||
```
Comment (peetucket) [inline app/api/sul_bib/authorship_api.rb]: so if I am reading this correctly, the updated process records class below will be called by harvester and take care of the discovery of existing WoS source records at publication creation time, right?

Comment (atz) [reply app/api/sul_bib/authorship_api.rb]: `author_uid` calls `author_contributions` and/or `process_records`.  The former queries `Publication` by `wos_uid` (and backfills contribs if needed).  `process_records` handles the case where nothing matches, or only a source record matches.

**Pull request merged**